src/prefect/utilities/threaded_heartbeat.py

```diff
@@ -0,0 +1,48 @@
+import threading
+from contextlib import contextmanager
+
+import prefect
+from prefect import config
+from prefect.client import Client
+from prefect.utilities.logging import get_logger
+
+
+class HeartbeatThread(threading.Thread):
+    def __init__(self, stop_event, flow_run_id, num=None):
+        threading.Thread.__init__(self)
+        self.num = num
+        self.daemon = True  # 'daemonizes' the thread, so Python will terminate it when all non-daemonized threads have finished
+        self.stop_event = stop_event
+
+    def run(self):
+        logger = get_logger('heartbeat')
+        client = Client()
+        iter_count = 0
+        with prefect.context({"flow_run_id": id, "running_with_backend": True}):
+            with log_heartbeat_failure(logger):
+                while iter_count < (self.num or 1) and not self.stop_event.is_set():
+                    self.stop_event.wait(timeout=config.cloud.heartbeat_interval)
+                    send_heartbeat(self.id, client, logger)
+                    iter_count += 1 if self.num else 0
+
+
+def send_heartbeat(id, client, logger):
+    try:  # Ignore (but log) client exceptions
+        client.update_flow_run_heartbeat(id)
+    except Exception as exc:
+        logger.error(
+            f"Failed to send heartbeat with exception: {exc!r}",
+            exc_info=True,
+        )
+
+
+@contextmanager
+def log_heartbeat_failure(logger):
```

Comment:
I think you can take this and `send_heartbeat` and move them into the executor utils. We should use these contexts from the flow run heartbeat CLI as well.